MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c

```diff
@@ -0,0 +1,1208 @@
+  IN  EFI_BLOCK_IO_PROTOCOL  *This,
+  IN  UINT32                 MediaId,
+  IN  EFI_LBA                Lba,
+  IN  UINTN                  BufferSize,
+  OUT VOID                   *Buffer
+  )
+{
+  NVME_DEVICE_PRIVATE_DATA  *Device;
+  EFI_STATUS                Status;
+  EFI_BLOCK_IO_MEDIA        *Media;
+  UINTN                     BlockSize;
+  UINTN                     NumberOfBlocks;
+  UINTN                     IoAlign;
+
+  //  EFI_TPL                   OldTpl;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeBlockIoReadBlocks: Enter\n"));
+
+  //
+  // Check parameters.
+  //
+  if (This == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Media = This->Media;
+
+  if (MediaId != Media->MediaId) {
+    return EFI_MEDIA_CHANGED;
+  }
+
+  if (Buffer == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (BufferSize == 0) {
+    return EFI_SUCCESS;
+  }
+
+  BlockSize = Media->BlockSize;
+  if ((BufferSize % BlockSize) != 0) {
+    return EFI_BAD_BUFFER_SIZE;
+  }
+
+  NumberOfBlocks = BufferSize / BlockSize;
+  if ((Lba + NumberOfBlocks - 1) > Media->LastBlock) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  IoAlign = Media->IoAlign;
+  if ((IoAlign > 0) && (((UINTN)Buffer & (IoAlign - 1)) != 0)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO (This);
+  Status = NvmeUnitTestRead (Device, Buffer, Lba, NumberOfBlocks);
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeBlockIoReadBlocks: Exit\n"));
+
+  // return EFI_SUCCESS;
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+NvmeBlockIoWriteBlocks (
+  IN  EFI_BLOCK_IO_PROTOCOL  *This,
+  IN  UINT32                 MediaId,
+  IN  EFI_LBA                Lba,
+  IN  UINTN                  BufferSize,
+  IN  VOID                   *Buffer
+  )
+{
+  NVME_DEVICE_PRIVATE_DATA  *Device;
+  EFI_STATUS                Status;
+  EFI_BLOCK_IO_MEDIA        *Media;
+  UINTN                     BlockSize;
+  UINTN                     NumberOfBlocks;
+  UINTN                     IoAlign;
+
+  //  EFI_TPL                   OldTpl;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeBlockIoWriteBlocks: Enter\n"));
+
+  //
+  // Check parameters.
+  //
+  if (This == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Media = This->Media;
+
+  if (MediaId != Media->MediaId) {
+    return EFI_MEDIA_CHANGED;
+  }
+
+  if (Buffer == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (BufferSize == 0) {
+    return EFI_SUCCESS;
+  }
+
+  BlockSize = Media->BlockSize;
+  if ((BufferSize % BlockSize) != 0) {
+    return EFI_BAD_BUFFER_SIZE;
+  }
+
+  NumberOfBlocks = BufferSize / BlockSize;
+  if ((Lba + NumberOfBlocks - 1) > Media->LastBlock) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  IoAlign = Media->IoAlign;
+  if ((IoAlign > 0) && (((UINTN)Buffer & (IoAlign - 1)) != 0)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO (This);
+  Status = NvmeUnitTestWrite (Device, Buffer, Lba, NumberOfBlocks);

```

Comment:
For now, that is correct. As more unit test cases get added there may be a need to examine the data itself that is passed (i.e. pattern recognition, buffer length, etc.). I would put this in the same category as the comment above. It can be addressed in a future PR to build more robustness to the unit tests.

Reply:
Yes, agreed. Let's make that change in a separate PR.